### ISSUE TITLE
Example to use word2vec with keras Convolution2D NN to do text classification?

### ISSUE THREAD

**Issue by eromoe** [fchollet/keras#5057]
There is Convolution1D example https://github.com/fchollet/keras/blob/master/examples/imdb_cnn.py without word2vec.

Current I am using gensim to train word2vec model.

I want to use word2vec and keras cnn to do document classifacation(Chinese Text ). I learned the basic flow of text classification in cnn and want to do a test.

## For example(the steps I imagine):

1. Use a good Cinese Tokenized Text Set to train word2vec model

        model = gensim.models.Word2Vec(new_sentences, workers=10, size=200, min_count=2)

2. Tokenize my sentences dataset to words lists dataset(the longest sentence has over 8000 words, shortest is less 50)
        
        1     ['你们', '好', '今天', '天气', '真', '好']
        2     ['嗯', '对的']
        ...
        9999  ['好', '就', '这样']

3. Use a method to transform words lists dataset to word2vec dataset

     transform every word in every sencence to a vec by trained model.

        1     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
        2     [[word2vec size=200], [word2vec size=200]]
        ...
        9999  [[word2vec size=200], [word2vec size=200], [word2vec size=200]]

4. Pad  word2vec dataset (with size=200 zero array)

        1     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
        2     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
        ....
        9999  [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]

5. go to the  CNN (using Convolution2D)

-------

I search for a long time, but can't find any way to do the step 3 ( after step 3,  the parameter and layers setting in step 5 is hard to understand too). I am not limit to gensim word2vec , could keras provide a full example?

**Comment by leocnj** [comment]
In your case, you can store the learned word2vec embedding to a text file. Then check [this example from Keras blog](https://blog.keras.io/using-pre-trained-word-embeddings-in-a-keras-model.html) to know how to load a pre-trained embedding.

**Comment by eromoe** [comment]
Thanks @leocnj .

Yestoday, I tried to construct X_train by below method:

```
def pad(array, reference_shape, offsets=None):
    """
    array: Array to be padded
    reference_shape: tuple of size of narray to create
    offsets: list of offsets (number of elements must be equal to the dimension of the array)
    will throw a ValueError if offsets is too big and the reference_shape cannot handle the offsets
    """
    
    if not offsets:
        offsets = np.zeros(array.ndim, dtype=np.float32)

    # Create an array of zeros with the reference shape
    result = np.zeros(reference_shape, dtype=np.float32)
    # Create a list of slices from offset to offset + shape in each dimension
    insertHere = [slice(offsets[dim], offsets[dim] + array.shape[dim]) for dim in range(array.ndim)]
    # Insert the array in the result at the specified offsets
    result[insertHere] = array
    return result


def pad_w2v(X_w2v, max_length):
    total = len(X_w2v)
    for i in xrange(total):
        X_w2v[i] = pad(X_w2v[i].astype(np.float32), (max_length, 200))
        
    return np.array(X_w2v)
```

Above method brought me  X_train like

```
 1     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
 2     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
 ....
 9999  [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
```

But it was very easy to get MemeryError,  finally I success to use (5800, 6810, 200) X_train with 1D conv  CNN , but the result is not good and eat up all ram 32G .... can't do anythin more. 

1. Is that mean I have to play with keras Embedding layer? Or is by my dataset  ? 

2. I always think 2D conv need a 3D input,  but I found actually 3D input only work with 1D conv, 2D conv need 4D input with shape like `(dataset_total_samples, 1, each_sample_feature_count, word_vector_size)` , is that right? 

3.  I am wondering if I did wrong work that transform every sentence to a list of word vector as below
```
 1     [[word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200], [word2vec size=200]]
 2     [[word2vec size=200], [word2vec size=200]]
 ...
 9999  [[word2vec size=200], [word2vec size=200], [word2vec size=200]]
```

Because Keras blog says:
> An Embedding layer should be fed sequences of integers, i.e. a 2D input of shape (samples, indices)

And this https://github.com/fchollet/keras/issues/233#issuecomment-117427013
> For each text input I break it up into it's constituent words, retrieve pretrained 300-dimensional word embeddings for these, and then concatenate them all, zero padding as needed to get a vector of a fixed length (30000).

**Comment by leocnj** [comment]
You can check this Github repo [A Keras implementation of CNN text classification following Yoon Kim EMNLP 2014](https://github.com/shagunsodhani/CNN-Sentence-Classifier) to directly learn the right way to use pre-trained embedding inside 1D Conv.

**Comment by harithaakella** [comment]
Hi, Could you actually share you code? It would help me with my problem which is very similar to yours. or send me the link to code on github.

**Comment by shaktisd** [comment]
In the examples given in https://github.com/shagunsodhani/CNN-Sentence-Classifier, it is using Glove , do you have a gensim implementation of the same ?
